src/DomainServices/DocumentService.ts

```diff
@@ -74,4 +101,145 @@ export class DocumentService {
+    const { manuscriptID, projectID } = validateListenUrl(url)
+
+    if (!manuscriptID || !projectID || !authToken) {
+      this.destroyStream(socket)
+      return
+    }
+    try {
+      this.validateTokenAccess(authToken, projectID, DocumentPermission.READ)
+    } catch (error) {
+      this.destroyStream(socket)
+      return
+    }
+
+    wss.handleUpgrade(request, socket, head, async (ws) =>
+      this.setupConnection(wss, socket, ws, request, manuscriptID)
+    )
+  }
+  private async setupConnection(
+    wss: WebSocketServer,
+    stream: Duplex,
+    ws: WebSocket,
+    request: IncomingMessage,
+    manuscriptID: string
+  ) {
+    wss.emit('connection', ws, request)
+    this.socketsService.setClient(manuscriptID, ws)
+    this.attachListeners(stream, ws, manuscriptID)
+    this.socketsService.send(JSON.stringify({ 'Transformer-Version': getVersion() }), ws)
+    log.info(`connection established for ${manuscriptID}`)
+  }
+
+  private attachListeners(stream: Duplex, ws: WebSocket, manuscriptID: string) {
+    ws.onerror = (error) => this.onError(ws, error)
+    ws.onclose = () => this.onClose(ws, manuscriptID)
+    ws.onmessage = (message) => this.onMessage(stream, ws, message)
+  }
+
+  private onError(ws: WebSocket, error: ErrorEvent) {
+    log.error(`webSocket error: ${error}`)
+    this.closeSocket(ws)
+  }
+  private onClose = (ws: WebSocket, manuscriptID: string) => {
+    log.info(`connection closed for ${manuscriptID}`)
+    this.closeSocket(ws, manuscriptID)
+  }
+
+  private async receiveSteps(
+    projectID: string,
+    manuscriptID: string,
+    payload: ReceiveSteps,
+    token: string
+  ) {
+    await this.validateTokenAccess(token, projectID, DocumentPermission.WRITE)
+    const history = await this.authorityService.receiveSteps(manuscriptID, payload)
+    this.socketsService.broadcast(manuscriptID, JSON.stringify(history))
+  }
+
+  private async onMessage(stream: Duplex, ws: WebSocket, message: MessageEvent) {
+    const { manuscriptID, projectID, authToken, payload } = this.parseMessage(message)
+    if (
+      !manuscriptID ||
+      !projectID ||
+      !authToken ||
+      !payload ||
+      (payload && !this.isReceiveStepsPayload(payload))
+    ) {
+      return this.handleInvalidData(stream, ws, manuscriptID)
+    }
+    try {
+      await this.receiveSteps(projectID, manuscriptID, payload, authToken)
+    } catch (error) {
+      if (error instanceof VersionMismatchError) {
+        this.socketsService.send(
+          JSON.stringify({ code: error.statusCode, error: error.message }),
+          ws
+        )
+      } else {
+        this.handleInvalidData(stream, ws, manuscriptID)
+        log.error(`error handling message: ${error}`)
+      }
+    }
+  }
+
+  private isReceiveStepsPayload(payload: any): payload is ReceiveSteps {
+    return (
+      Array.isArray(payload.steps) &&
+      payload.steps.every((step: object) => typeof step === 'object' && step !== null) &&
```

Comment:
Do we need to go into this much detail here? Is the array check not sufficient?

Reply:
Yeah it should be sufficient.